### ISSUE TITLE
Refactor Expression.py and runExpression.py

### ISSUE THREAD

**Issue by TomSmithCGAT** [CGATOxford/cgat#130]
Currently, runExpression.py uses the Expression.py module to run differential expression tools including edgeR, DESeq as well as summarising counts tables and differential expression tables. 

Expression.py contains many functions which rely upon r objects existing in the r namespace. For example, to run DESeq, loadTagData is called to create the following objects in the r namespace: countsTable, groups, conds and pairs. Then filterTagData is called which relies upon these objects existing in the r namespace to filter the countsTable. Then runDESeq is called which again relies upon pre-existing objects in the r namespace. 

A more sensible approach would be to perform the load and filter steps in pandas using the Counts.py module and then pass counts and design dataframes along with other variables to the runDESeq function. 

There are also a number of improvements that could be made to Expression.py, inlcuding the addition of DESeq2 which is currently being implemented by Jethro. 

The proposed changes are:

1. Implement DESeq2 (Expression.py / runExpression.py)
2. Perform all loading and filtering steps using pandas  (Counts.py)
3. Recode edgeR/DESeq etc functions to accept pandas dataframes and associated variables as inputs (Expression.py)
4. Implement normalisation pre-percentile filtering but return a table of filtered raw counts (Expression.py)
6. Refactor calls to plotting functions into a counts2summary.py script (Expression.py / counts2summary.py / runExpression.py)

Please feel free to add any more suggestions you have.

We're going to work on a branch called "JJTS-RefactorExpressionAndRunExpression"

**Comment by AndreasHeger** [comment]
Thanks, that is a great project!

**Comment by TomSmithCGAT** [comment]
After a discussion with Jethro, we've decided to make a class DEcaller in Expression.py with instances for each differential expression caller (edger, DESeq, DESeq2, TTest).

A pandas DataFrame of counts will be passed to the DECaller, along with the design file and parameter such as the reference name, model, contrasts etc.

The methods of the class will perform the following:
1. Parse the design file
# identify groups, whether there are replicates, etc

2. Filter the counts DF
# steps 1 & 2 will be defined in the Parent class to use functions in Class.py module file and shouldn't need overriding in child classes

3. Perform differential expression
# This will mainly involve the implementation of R code from the respective vignettes using rpy2
to include as many of the analysis methods as possible (e.g no replicates, paired, glm followed by contrasts)

4. Make diagnostic plots
# For each tool there are standard plots the user will expect to see according to the vignette. This method will generate these plots.

5. Post-process results
# This step will process the results into a final results table including a BH-adjusted p-value across all tests performed

We envisage that additional steps such as summarising the number of significant DE genes will be performed in a subsequent step in the pipeline. Hence this won't be included in the class methods. 

Likewise, we think it would be helpful to generate standardised plots for comparison across tools such as MA plots. Again, these would be generated outside of the class methods.

**Comment by AndreasHeger** [comment]
Thanks, these are good suggestions, my thoughts are to separate parsing/algorithm/output/reporting.
This means, the DE class should accept the design (a Design class/a data structure), the model and the counts and return a standardized set of results. Maybe provide provisions for method specific QA plots and metrics.
Parsing a design table, standardized plotting, etc should be done in auxilliary classes or functions.

**Comment by TomSmithCGAT** [comment]
I agree this would make the code more reuable and clearer. 

I've recoded the loadTagData and filterTagData functions in counts.py to methods for a "Counts" class. I've also made a "Design" class in Expression.py with associated methods to get attributes (e.g "has_replicates") from the design table, and a "DEExperiment" class in Expression.py to instantiate, modify and hold the Counts and Design objects for a particular experiment. 

My idea would be to then pass a DEExperiment object along with a model etc to the DECaller class.
